Regression of downstream builds due to an exported target that's not found

On May 22nd a new release was made which caused a regression in the Fbin builds on all platforms. 

There are 3 core packages now failing to build due to pthread detection failures in CMake linking against pthreads.


For example: http://build.ros2.org/view/Fbin_uF64/job/Fbin_uF64__tf2__ubuntu_focal_amd64__binary/22/console

<detail>

```
DCMAKE_HAVE_LIBC_PTHREAD    CMakeFiles/cmTC_a9f57.dir/src.c.o  -o cmTC_a9f57 
10:59:53 /usr/bin/ld: CMakeFiles/cmTC_a9f57.dir/src.c.o: in function `main':
10:59:53 ./obj-x86_64-linux-gnu/CMakeFiles/CMakeTmp/./obj-x86_64-linux-gnu/CMakeFiles/CMakeTmp/src.c:11: undefined reference to `pthread_create'
10:59:53 /usr/bin/ld: ./obj-x86_64-linux-gnu/CMakeFiles/CMakeTmp/./obj-x86_64-linux-gnu/CMakeFiles/CMakeTmp/src.c:12: undefined reference to `pthread_detach'
10:59:53 /usr/bin/ld: ./obj-x86_64-linux-gnu/CMakeFiles/CMakeTmp/./obj-x86_64-linux-gnu/CMakeFiles/CMakeTmp/src.c:13: undefined reference to `pthread_join'
10:59:53 collect2: error: ld returned 1 exit status
10:59:53 make[3]: *** [CMakeFiles/cmTC_a9f57.dir/build.make:87: cmTC_a9f57] Error 1
10:59:53 make[3]: Leaving directory '/tmp/binarydeb/ros-foxy-tf2-0.13.2/obj-x86_64-linux-gnu/CMakeFiles/CMakeTmp'
10:59:53 make[2]: *** [Makefile:121: cmTC_a9f57/fast] Error 2
10:59:53 make[2]: Leaving directory '/tmp/binarydeb/ros-foxy-tf2-0.13.2/obj-x86_64-linux-gnu/CMakeFiles/CMakeTmp'
10:59:53 
10:59:53 
10:59:53 Source file was:
10:59:53 #include <pthread.h>
10:59:53 
10:59:53 void* test_func(void* data)
10:59:53 {
10:59:53   return data;
10:59:53 }
10:59:53 
10:59:53 int main(void)
10:59:53 {
10:59:53   pthread_t thread;
10:59:53   pthread_create(&thread, NULL, test_func, NULL);
10:59:53   pthread_detach(thread);
10:59:53   pthread_join(thread, NULL);
10:59:53   pthread_atfork(NULL, NULL, NULL);
10:59:53   pthread_exit(NULL);
10:59:53 
10:59:53   return 0;
10:59:53 }
10:59:53 
10:59:53 make[1]: Leaving directory '/tmp/binarydeb/ros-foxy-tf2-0.13.2'
10:59:53 dh_auto_configure: error: cd obj-x86_64-linux-gnu && cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON "-GUnix Makefiles" -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_AUTOGEN_VERBOSE=ON -DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu -DCMAKE_INSTALL_PREFIX=/opt/ros/foxy -DAMENT_PREFIX_PATH=/opt/ros/foxy .. returned exit code 1
10:59:53 make[1]: *** [debian/rules:28: override_dh_auto_configure] Error 25
10:59:53 make: *** [debian/rules:22: build] Error 2
10:59:53 dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
```
</detail>


The first failed build appears to be 

- tf2: http://build.ros2.org/view/Fbin_uF64/job/Fbin_uF64__tf2__ubuntu_focal_amd64__binary/22/
- rmw_fastrtps_shared_cpp: http://build.ros2.org/view/Fbin_uF64/job/Fbin_uF64__rmw_fastrtps_shared_cpp__ubuntu_focal_amd64__binary/18/
- rmw_cyclonedds_cpp: http://build.ros2.org/view/Fbin_uF64/job/Fbin_uF64__rmw_cyclonedds_cpp__ubuntu_focal_amd64__binary/19/

Each of those successively has the following new releases triggered. 


```
tfoote@snowman4:/tmp/root_cause Last: [0] (0s Seconds)
$ cat tf2.triggers | grep Fsrc | awk '{print $5}' | sort > tf2.new
tfoote@snowman4:/tmp/root_cause Last: [0] (0s Seconds)
$ cat cyclonedds_cpp.triggers | grep Fsrc | awk '{print $5}' | sort > cyclone.new
tfoote@snowman4:/tmp/root_cause Last: [0] (0s Seconds)
$ cat fastrtps.triggers | grep Fsrc | awk '{print $5}' | sort > fastrtps.new
tfoote@snowman4:/tmp/root_cause Last: [0] (0s Seconds)
$ meld *.new
```

![Screenshot from 2020-05-22 15-48-19](https://user-images.githubusercontent.com/447804/82714816-34c19600-9c45-11ea-8499-f2b8cf571cae.png)


```
tfoote@snowman4:/tmp/root_cause Last: [0] (46s Seconds)
$ cat tf2.new
Fsrc_uF__fastrtps_cmake_module__ubuntu_focal__source
Fsrc_uF__osrf_testing_tools_cpp__ubuntu_focal__source
Fsrc_uF__rcutils__ubuntu_focal__source
Fsrc_uF__rosidl_adapter__ubuntu_focal__source
Fsrc_uF__rosidl_cmake__ubuntu_focal__source
Fsrc_uF__rosidl_generator_cpp__ubuntu_focal__source
Fsrc_uF__rosidl_generator_c__ubuntu_focal__source
Fsrc_uF__rosidl_parser__ubuntu_focal__source
Fsrc_uF__rosidl_runtime_cpp__ubuntu_focal__source
Fsrc_uF__rosidl_runtime_c__ubuntu_focal__source
Fsrc_uF__rosidl_typesupport_fastrtps_cpp__ubuntu_focal__source
Fsrc_uF__rosidl_typesupport_fastrtps_c__ubuntu_focal__source
Fsrc_uF__rosidl_typesupport_interface__ubuntu_focal__source
Fsrc_uF__rosidl_typesupport_introspection_cpp__ubuntu_focal__source
Fsrc_uF__rosidl_typesupport_introspection_c__ubuntu_focal__source
```


Looking a little higher in the logs http://build.ros2.org/view/Fbin_uF64/job/Fbin_uF64__rmw_cyclonedds_cpp__ubuntu_focal_amd64__binary/19/console

I found


<details>

```
10:45:24 CMake Error at CMakeLists.txt:55 (add_library):
10:45:24   Target "rmw_cyclonedds_cpp" links to target
10:45:24   "rosidl_typesupport_fastrtps_cpp::rosidl_typesupport_fastrtps_cpp" but the
10:45:24   target was not found.  Perhaps a find_package() call is missing for an
10:45:24   IMPORTED target, or an ALIAS target is missing?
10:45:24 
10:45:24 
10:45:24 CMake Error at CMakeLists.txt:55 (add_library):
10:45:24   Target "rmw_cyclonedds_cpp" links to target
10:45:24   "rosidl_typesupport_fastrtps_cpp::rosidl_typesupport_fastrtps_cpp" but the
10:45:24   target was not found.  Perhaps a find_package() call is missing for an
10:45:24   IMPORTED target, or an ALIAS target is missing?
10:45:24 
10:45:24 
10:45:24 CMake Error at CMakeLists.txt:55 (add_library):
10:45:24   Target "rmw_cyclonedds_cpp" links to target
10:45:24   "rosidl_typesupport_fastrtps_c::rosidl_typesupport_fastrtps_c" but the
10:45:24   target was not found.  Perhaps a find_package() call is missing for an
10:45:24   IMPORTED target, or an ALIAS target is missing?
10:45:24 
10:45:24 
10:45:24 -- Generating done
10:45:24 CMake Warning:
10:45:24   Manually-specified variables were not used by the project:
10:45:24 
10:45:24     AMENT_PREFIX_PATH
10:45:24     CMAKE_AUTOGEN_VERBOSE
10:45:24     CMAKE_EXPORT_NO_PACKAGE_REGISTRY
10:45:24     CMAKE_INSTALL_LIBDIR
10:45:24     CMAKE_INSTALL_LOCALSTATEDIR
10:45:24     CMAKE_INSTALL_SYSCONFDIR
10:45:24 
10:45:24 
10:45:24 CMake Generate step failed.  Build files cannot be regenerated correctly.
10:45:24 	cd obj-x86_64-linux-gnu && tail -v -n \+0 CMakeCache.txt
```
</details>

Also in http://build.ros2.org/view/Fbin_uF64/job/Fbin_uF64__tf2__ubuntu_focal_amd64__binary/22/consoleFull
and http://build.ros2.org/view/Fbin_uF64/job/Fbin_uF64__rmw_fastrtps_shared_cpp__ubuntu_focal_amd64__binary/18/console

I believe that it's an issue with the 1.0 release of rosidl_typesupport_fastrtps_cpp http://build.ros2.org/job/Fsrc_uF__rosidl_typesupport_fastrtps_cpp__ubuntu_focal__source/2/

The most recent change in a related code path appears to be https://github.com/ros2/rosidl_typesupport_fastrtps/pull/40 but I'm not familiar with those changes say anything beyond that's where I'd look first.

--- a/rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake
+++ b/rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c_generate_interfaces.cmake
@@ -12,12 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-if(NOT rosidl_generator_c_FOUND)
-  message(FATAL_ERROR
-    "'rosidl_generator_c' not found when executing "
-    "'rosidl_typesupport_fastrtps_c' extension.")
-endif()
-
 find_package(ament_cmake_ros REQUIRED)
 find_package(fastrtps_cmake_module QUIET)
 find_package(fastcdr REQUIRED CONFIG)
@@ -129,10 +123,15 @@ set_target_properties(${rosidl_generate_interfaces_TARGET}${_target_suffix}
   PROPERTIES COMPILE_FLAGS "${_target_compile_flags}")
 target_include_directories(${rosidl_generate_interfaces_TARGET}${_target_suffix}
   PUBLIC
-  "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_c>"
-  "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_cpp>"
-  "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/rosidl_typesupport_fastrtps_c>"
-  "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/rosidl_typesupport_fastrtps_cpp>"
+  ${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_c
+  ${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_cpp
+  ${CMAKE_CURRENT_BINARY_DIR}/rosidl_typesupport_fastrtps_c
+  ${CMAKE_CURRENT_BINARY_DIR}/rosidl_typesupport_fastrtps_cpp
+)
+ament_target_dependencies(${rosidl_generate_interfaces_TARGET}${_target_suffix}
+  "fastrtps"
+  "rosidl_typesupport_fastrtps_cpp"
+  "rosidl_typesupport_fastrtps_c"
 )
 foreach(_pkg_name ${rosidl_generate_interfaces_DEPENDENCY_PACKAGE_NAMES})
   set(_msg_include_dir "${${_pkg_name}_DIR}/../../../include/${_pkg_name}/msg/dds_fastrtps_c")
@@ -143,14 +142,14 @@ foreach(_pkg_name ${rosidl_generate_interfaces_DEPENDENCY_PACKAGE_NAMES})
   normalize_path(_action_include_dir "${_action_include_dir}")
   target_include_directories(${rosidl_generate_interfaces_TARGET}${_target_suffix}
     PUBLIC
-    "$<BUILD_INTERFACE:${_msg_include_dir}>"
-    "$<BUILD_INTERFACE:${_srv_include_dir}>"
-    "$<BUILD_INTERFACE:${_action_include_dir}>"
+    "${_msg_include_dir}"
+    "${_srv_include_dir}"
+    "${_action_include_dir}"
   )
   ament_target_dependencies(${rosidl_generate_interfaces_TARGET}${_target_suffix}
     ${_pkg_name})
   target_link_libraries(${rosidl_generate_interfaces_TARGET}${_target_suffix}
-    ${${_pkg_name}_TARGETS${_target_suffix}})
+    ${${_pkg_name}_LIBRARIES${_target_suffix}})
 endforeach()
 target_link_libraries(${rosidl_generate_interfaces_TARGET}${_target_suffix}
   ${rosidl_generate_interfaces_TARGET}__rosidl_generator_c
@@ -182,14 +181,13 @@ if(NOT rosidl_generate_interfaces_SKIP_INSTALL)
 
   install(
     TARGETS ${rosidl_generate_interfaces_TARGET}${_target_suffix}
-    EXPORT ${rosidl_generate_interfaces_TARGET}${_target_suffix}
     ARCHIVE DESTINATION lib
     LIBRARY DESTINATION lib
     RUNTIME DESTINATION bin
   )
-  rosidl_export_typesupport_targets(${_target_suffix}
+
+  rosidl_export_typesupport_libraries(${_target_suffix}
     ${rosidl_generate_interfaces_TARGET}${_target_suffix})
-  ament_export_targets(${rosidl_generate_interfaces_TARGET}${_target_suffix})
 endif()
 
 if(BUILD_TESTING AND rosidl_generate_interfaces_ADD_LINTER_TESTS)

--- a/rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c-extras.cmake.in
+++ b/rosidl_typesupport_fastrtps_c/rosidl_typesupport_fastrtps_c-extras.cmake.in
@@ -11,10 +11,6 @@ if(NOT FastRTPS_FOUND)
     "Could not find eProsima Fast-RTPS: skipping rosidl_typesupport_fastrtps_c"
   )
 else()
-  # Make sure extension points are registered in order
-  find_package(rosidl_generator_c QUIET)
-  find_package(rosidl_typesupport_fastrtps_cpp QUIET REQUIRED)
-
   find_package(ament_cmake_core QUIET REQUIRED)
   ament_register_extension(
     "rosidl_generate_idl_interfaces"

--- a/rosidl_typesupport_fastrtps_cpp/cmake/rosidl_typesupport_fastrtps_cpp_generate_interfaces.cmake
+++ b/rosidl_typesupport_fastrtps_cpp/cmake/rosidl_typesupport_fastrtps_cpp_generate_interfaces.cmake
@@ -12,12 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-if(NOT rosidl_generator_cpp_FOUND)
-  message(FATAL_ERROR
-    "'rosidl_generator_cpp' not found when executing "
-    "'rosidl_typesupport_fastrtps_cpp' extension.")
-endif()
-
 find_package(ament_cmake_ros REQUIRED)
 find_package(fastrtps_cmake_module QUIET)
 find_package(fastcdr REQUIRED CONFIG)
@@ -142,8 +136,8 @@ set_target_properties(${rosidl_generate_interfaces_TARGET}${_target_suffix}
 # Include headers from other generators
 target_include_directories(${rosidl_generate_interfaces_TARGET}${_target_suffix}
   PUBLIC
-  "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_cpp>"
-  "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/rosidl_typesupport_fastrtps_cpp>"
+  ${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_cpp
+  ${CMAKE_CURRENT_BINARY_DIR}/rosidl_typesupport_fastrtps_cpp
 )
 
 ament_target_dependencies(${rosidl_generate_interfaces_TARGET}${_target_suffix}
@@ -158,7 +152,7 @@ foreach(_pkg_name ${rosidl_generate_interfaces_DEPENDENCY_PACKAGE_NAMES})
   ament_target_dependencies(${rosidl_generate_interfaces_TARGET}${_target_suffix}
     ${_pkg_name})
   target_link_libraries(${rosidl_generate_interfaces_TARGET}${_target_suffix}
-    ${${_pkg_name}_TARGETS${_target_suffix}})
+    ${${_pkg_name}_LIBRARIES${_target_suffix}})
 endforeach()
 
 target_link_libraries(${rosidl_generate_interfaces_TARGET}${_target_suffix}
@@ -190,14 +184,13 @@ if(NOT rosidl_generate_interfaces_SKIP_INSTALL)
 
   install(
     TARGETS ${rosidl_generate_interfaces_TARGET}${_target_suffix}
-    EXPORT ${rosidl_generate_interfaces_TARGET}${_target_suffix}
     ARCHIVE DESTINATION lib
     LIBRARY DESTINATION lib
     RUNTIME DESTINATION bin
   )
-  rosidl_export_typesupport_targets(${_target_suffix}
+
+  rosidl_export_typesupport_libraries(${_target_suffix}
     ${rosidl_generate_interfaces_TARGET}${_target_suffix})
-  ament_export_targets(${rosidl_generate_interfaces_TARGET}${_target_suffix})
 endif()
 
 if(BUILD_TESTING AND rosidl_generate_interfaces_ADD_LINTER_TESTS)

--- a/rosidl_typesupport_fastrtps_cpp/rosidl_typesupport_fastrtps_cpp-extras.cmake.in
+++ b/rosidl_typesupport_fastrtps_cpp/rosidl_typesupport_fastrtps_cpp-extras.cmake.in
@@ -12,9 +12,6 @@ if(NOT FastRTPS_FOUND)
     "Could not find eProsima Fast-RTPS - skip rosidl_typesupport_fastrtps_cpp"
   )
 else()
-  # Make sure extension points are registered in order
-  find_package(rosidl_generator_cpp QUIET)
-
   find_package(ament_cmake_core QUIET REQUIRED)
   ament_register_extension(
     "rosidl_generate_idl_interfaces"

